packaging/metadata/_validation.py

```diff
@@ -0,0 +1,159 @@
+from __future__ import annotations
+
+import abc
+import re
+from typing import TYPE_CHECKING, Any, Callable, Dict, Generic, Optional, TypeVar, cast
+
+if TYPE_CHECKING:
+    from ._types import Metadata
+
+
+T = TypeVar("T")
+
+
+class lazy_validator(Generic[T]):  # noqa: N801
+
+    # This hack exists to work around https://github.com/python/mypy/issues/708
+    _creator: Callable[[Any], T] | Callable[[Any], T]
+    _raw_name: str
+    _validators: list[Callable[[Any], None]]
+
+    def __init__(
+        self,
+        creator: Callable[[Any], T],
+        *,
+        raw_name: str | None = None,
+        validators: list[Callable[[Any], None]] | None = None,
+    ) -> None:
+        self._creator = creator
+        if raw_name is not None:
+            self._raw_name = raw_name
```

Comment:
Instead of a `hasattr()` check one can just use an `if` check in `__set_name__`, else it looks suspicious as to why `self._raw_name` can be left unset.

```suggestion
        self._raw_name = raw_name
```